### PR TITLE
gdb: enable to specify multiple candidates for gdb

### LINKED ISSUE #1662: How to do tinygo-gdb on windows
The following changes make me use gdb-multiarch, but I still think I need to use arm-none-eabi-gdb.exe on windows.

* 1a2f051b

If rename arm-none-eabi-gdb.exe to gdb-multiarch.exe, it can be debugged with tinygo-gdb.
However, I think it is better to specify it with an argument like `tinygo gdb -debugger=arm-none-eabi-gdb ...` rather than renaming it.
Since I already created something that works, I will create a Pull-Request if this approach is okay.

or,,,
Is there any other better way?

windows 10 + tinygo 0.16 or later + cortex-m4

```
error: failed to run gdb with C:\Users\sago3\AppData\Local\Temp\tinygo160016967\main: exec: "gdb-multiarch": executable file not found in
 %PATH%
```

arm-neno-eabi-gdb for windows can be found at:
https://developer.arm.com/tools-and-software/open-source-software/developer-tools/gnu-toolchain/gnu-rm/downloads

### PR DESCRIPTION
**Pull request by sago35**
fixes #1662 (How to do tinygo-gdb on windows)

### CODE REVIEW

File: compileopts/target.go [hunk 1]
```diff
@@ -271,3 +272,16 @@ func defaultTarget(goos, goarch, triple string) (*TargetSpec, error) {
 	}
 	return &spec, nil
 }
+
+func (spec *TargetSpec) Debugger() (string, error) {
```
Comment (aykevl) [inline compileopts/target.go]: Can you rename this to `GDB`?

```suggestion
func (spec *TargetSpec) GDB() (string, error) {
```

We might want to add support for other debuggers in the future, such as LLDB (which can possibly be included in the TinyGo release, unlike GDB). LLDB will need special support, it is not a drop-in replacement.

File: main.go [hunk 2]
```diff
@@ -484,11 +485,11 @@ func FlashGDB(pkgName string, ocdOutput bool, options *compileopts.Options) erro
 		for _, cmd := range gdbCommands {
 			params = append(params, "-ex", cmd)
 		}
-		cmd := executeCommand(config.Options, config.Target.GDB, params...)
+		cmd := executeCommand(config.Options, gdb, params...)
 		cmd.Stdin = os.Stdin
 		cmd.Stdout = os.Stdout
 		cmd.Stderr = os.Stderr
-		err := cmd.Run()
+		err = cmd.Run()
```
Comment (aykevl) [inline main.go]: Was this intentional?

File: compileopts/target.go (hunk 1 above)
Comment (sago35) [reply compileopts/target.go]: If I rename it to GDB(), I get the following error.
Is there any other name that would be better?

```
compileopts\target.go:276:6: type TargetSpec has both field and method named GDB
```

File: main.go (hunk 2 above)
Comment (sago35) [reply main.go]: It was an unnecessary change.
Fixed.

File: compileopts/target.go (hunk 1 above)
Comment (aykevl) [reply compileopts/target.go]: Ah, I see. I can't think of a good name, maybe `GDBCommand` or `LookupGDB`?

Comment (sago35) [reply compileopts/target.go]: I renamed Debugger() to LookupGDB().

File: compileopts/target.go [hunk 3]
```diff
@@ -253,7 +254,7 @@ func defaultTarget(goos, goarch, triple string) (*TargetSpec, error) {
 	}
 	if goarch != runtime.GOARCH {
 		// Some educated guesses as to how to invoke helper programs.
-		spec.GDB = "gdb-multiarch"
+		spec.GDB = []string{"gdb-multiarch", "arm-none-eabi-gdb"}
```
Comment (aykevl) [inline compileopts/target.go]: This change is not correct. This code is for regular desktop usage, where `gdb-multiarch` is usually the correct name and `arm-none-eabi-gdb` never is.

File: compileopts/target.go [hunk 4]
```diff
@@ -271,3 +272,17 @@ func defaultTarget(goos, goarch, triple string) (*TargetSpec, error) {
...
+	if len(spec.GDB) == 0 {
+		return "", errors.New("gdb not configured in the target specification")
+	}
+	for _, d := range spec.GDB {
+		_, err := exec.LookPath(d)
+		if err == nil {
+			return d, nil
+		}
+	}
+	return "", errors.New("no gdb found configured in the target specification")
```
Comment (aykevl) [inline compileopts/target.go]: Can you maybe include the command names that were checked in the error message? That makes it easier to know how to fix the issue based on the error message.

File: targets/cortex-m7.json [hunk 5]
```diff
@@ -5,5 +5,5 @@
 		"--target=armv7em-none-eabi",
 		"-Qunused-arguments"
 	],
-	"gdb": "arm-none-eabi-gdb"
+	"gdb": ["arm-none-eabi-gdb"]
```
Comment (aykevl) [inline targets/cortex-m7.json]: Strictly speaking this line is not necessary, but it was already present before so it's fine (no change needed). But you can remove it if you want.

File: compileopts/target.go (hunk 4 above)
Comment (sago35) [reply compileopts/target.go]: Added information to error message.

```
error: no gdb found configured in the target specification (gdb-multiarch, arm-none-eabi-gdb)
```

File: compileopts/target.go (hunk 3 above)
Comment (sago35) [reply compileopts/target.go]: arm-none-eabi-gdb has been removed.

File: targets/cortex-m7.json (hunk 5 above)
Comment (sago35) [reply targets/cortex-m7.json]: Removed because it is defined in the inherited cortex-m.json

**Comment by sago35** [comment]
Please review this PR.
 @aykevl

**Comment by oflebbe** [comment]
This patch is needed for MacOS as well. +1

**Comment by deadprogram** [comment]
I think this PR seems OK to merge if squashed. @aykevl what do you think?

**Comment by deadprogram** [comment]
@sago35 can you please squash this PR so we can merge it?

@aykevl any further feedback?

**Comment by sago35** [comment]
I squashed and rebased it. @deadprogram

**Comment by deadprogram** [comment]
Thanks @sago35 now merging.

**Comment by sago35** [comment]
Thank you!

**Pull request merged**